### PR TITLE
Use Google resource key if available in the URL

### PR DESCRIPTION
**Pull request by jon-betts**
This also includes much fancier Google Drive URL parsing as we've seen the URLs have more variation in them than we previously thought.

Based on:

 * https://github.com/hypothesis/via/pull/583
 * https://github.com/hypothesis/via/pull/581
 * https://github.com/hypothesis/via/pull/580
 
Requires:

 * https://github.com/hypothesis/devdata/pull/31

## Testing notes

All of this should be done with caching diabled.

Either drop a debug point into `via/services/google_drive.py` or modify it to log when it adds a resource key:

```diff
--- a/via/services/google_drive.py
+++ b/via/services/google_drive.py
@@ -111,6 +111,9 @@ class GoogleDriveAPI:
             resource_key = self._resource_keys.get(file_id)
 
         if resource_key:
+            from logging import getLogger
+            getLogger(__name__).warning(f"Resource key {resource_key}")
+
             headers["X-Goog-Drive-Resource-Keys"] = f"{file_id}/{resource_key}"
 
         response = self._session.get(url=url, headers=headers, stream=True, timeout=10)
```

You should be able to see the resource key printed out when you visit:

 * http://0.0.0.0:9082/google_drive/0ByrDah1tVd5qMmVzeTF3a0dDOWc/AMADEUPRESOURCEKEY/proxied.pdf
 * The above demonstrates proxying the resource key
 * http://0.0.0.0:9082/google_drive/0ByrDah1tVd5qMmVzeTF3a0dDOWc/proxied.pdf
 * The above demonstrates looking up the value

We should extract the key from URLs. Try pasting these into the root:

 * https://drive.google.com/uc?id=0ByrDah1tVd5qMmVzeTF3a0dDOWc&export=download&resourceKey=AMADEUPRESOURCEKEY
* https://drive.google.com/uc?id=0ByrDah1tVd5qMmVzeTF3a0dDOWc&export=download&resourcekey=AMADEUPRESOURCEKEY

We also accept a wider range of URLs now:

 * https://drive.google.com/file/d/17zLSXuH87GhQJRqTXDxP-vGO--GkeaC8/view?usp=sharing
 * https://drive.google.com/uc?id=0BzxGAA_njUeBSE5wdHlxblJtck0

Try pasting these into the landing page and observe the mapped URL

### CODE REVIEW

**Comment by seanh** [comment]
@jon-betts Could you add manual testing instructions to this please?

**Review by marcospri — APPROVED** [review]
Looks good :+1: 

For the real effect on google enforcing the resource key param we'll have to wait until the 13th but I think we should merge it now.

**Comment by jon-betts** [comment]
Yes, unfortunately you can't really test it because Google Drive API seems to ignore the header for now. Or at least is indifferent to it being wrong.

File: via/views/view_pdf.py
Comment (seanh) [inline via/views/view_pdf.py:56]: :+1: Nice, this seems like a nice helper class

**Pull request merged**
